### PR TITLE
Custom Actions and Chuchu key improvements

### PR DESCRIPTION
**Pull request by bhoffman20**
Make the chuchu key more customizable by adding shortcuts to custom actions, and making built-in shortcuts editable.

### Features

- **Added builtin chuchu command shortcuts** (customizable via Settings → Chuchu Commands)
  - `a` — toggle floating actions button visibility
  - `s` — open settings
  - `q` — close active tab
- **Shortcut field on custom actions** — bind a single-character key to any custom action so it can be triggered from chuchu command mode
- **Multi-action shortcut disambiguation** — when a key maps to multiple custom actions, a filtered FAB opens to pick one; action names shown in the hint bar
- **Persistent actions FAB visibility toggle** in settings
- **Command panel rendered as a floating overlay** instead of resizing the terminal

### Refactoring

- **`BuiltinCommand` enum** — single source of truth for builtin commands (id, label, description), replacing inline hardcoded strings. Enum order drives both settings UI and hint bar ordering.
- **`BuiltinShortcutStore`** — JSON-backed persistence object for builtin shortcut key bindings, with `defaults`, `normalize`, `serialize`, and `parse` surface. Key collision detection forces one-key-per-command.
- **Custom actions storage migrated to JSON** — `TerminalCustomActionStore` now serializes as a JSON array, replacing the legacy delimited format (`key=label::payload|...`) which was vulnerable to delimiter collision. Legacy data is transparently migrated on load.
- **`TerminalCustomAction.shortcut` field** — new optional `String?` property added to the data class, serialized in JSON, with runtime lowercase normalization for consistent lookup.
- **Chuchu hint bar** — builtin hints are ordered by `BuiltinCommand.entries` iteration (deterministic, matching settings) instead of arbitrary map-key order. Custom action hints follow, grouped by shortcut key with multi-action labels.
- **Key normalization** — all shortcut keys are lowercased at parse/normalize time so that stored keys, settings conflict checks, hint labels, and runtime matching agree.
- **`TerminalCustomActionsEditor`** — extracted `cancelEdit()` helper, reordered fields (Key → Shortcut → Modifiers → Value), inline shortcut display in list rows, collapsing edit panel on re-click.
- **New tests** — `BuiltinShortcutStoreTest` and `TerminalCustomActionStoreTest` covering serialization round-trips, legacy format migration, normalization, and edge cases.

### Edge Case Handling

- **Multiple custom actions can be assigned the same shortcut key** – When multiple custom actions are assigned the same shortcut key, the pop-up shows `b: [action, other_action]` and opens up a filtered list on the custom actions FAB showing the actions assigned to that shortcut.
- **Built in commands always get the shortcut key** – If a shortcut key is assigned to a custom action *AND* a built in action (e.g. "n" is used by "new tab", and assigned to a custom action), the shortcut simply doesn't get used for the custom action at all. The user would have to reassign the "new tab" action to a different key (or remove it) before it could be used with a custom action.

### Screenshots
New settings added:
Screenshot_20260617_104750_chuchu

Custom Actions Editor:
Screenshot_20260617_105014_chuchu

Chuchu Key Popup:
Screenshot_20260617_105124_chuchu

NOTE: I did not change the default shortcut key for "new tab", I modified it via the new settings. The default key is still "n".

NOTE 2: Ignore the look of the Accessory keys bar in my screenshots, that is part of a different change I am working on. The default look and feel of the accessory keys has not changed in this PR.

## Summary by CodeRabbit

* **New Features**
  * Added settings to show/hide the custom-actions floating button and to customize built-in command shortcuts.
  * Added a built-in commands editor sheet, and integrated both the editor and shortcut state into the settings/terminal flows.
  * Enhanced custom actions with optional single-key shortcuts and improved terminal keybinding behavior (including filtered-actions FAB mode).
* **Bug Fixes**
  * Improved shortcut persistence by switching to JSON-first storage and consistently normalizing, deduplicating, and handling malformed data.
* **Tests**
  * Added/expanded tests for built-in and custom shortcut JSON parsing, normalization, and legacy migration.
* **Chores**
  * Enabled Kotlin serialization and added JSON runtime support.

### CODE REVIEW

**Comment by jossephus** [comment]
Hey ben, Thanks for the PR. You have so many good idea here. 
I will take a look on the weekend.

File: android/app/src/main/java/com/jossephus/chuchu/ui/terminal/BuiltinCommands.kt [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+ * Uses JSON (built-in org.json) so any key char — including `:`, `,`, `;`, quotes — is
+ * safe, mirroring [TerminalCustomActionStore]. The pref key is new on this branch, so
+ * there is no legacy delimited format to migrate.
+ */
+object BuiltinShortcutStore {
+    val defaults: Map<String, String> = mapOf(
+        BuiltinCommand.Tabs.id to "t",
+        BuiltinCommand.NewTab.id to "n",
+        BuiltinCommand.Actions.id to "a",
+        BuiltinCommand.Settings.id to "s",
```
Comment (jossephus) [inline android/app/src/main/java/com/jossephus/chuchu/ui/terminal/BuiltinCommands.kt]: make the default key part of the BuiltinCommand enum.
```
    NewTab("new_tab", "new tab", "open a new tab", 'n'),
```

File: android/app/src/main/java/com/jossephus/chuchu/ui/terminal/TerminalCustomActions.kt [hunk 2]
```diff
@@ -99,7 +105,38 @@ object TerminalCustomActionStore {
     fun parse(raw: String?): List<TerminalCustomKeyGroup> {
         if (raw == null) return defaultGroups
         if (raw.isBlank()) return emptyList()
-        val parsed = raw
+        val parsed = try {
+            val array = JSONArray(raw)
```
Comment (jossephus) [inline android/app/src/main/java/com/jossephus/chuchu/ui/terminal/TerminalCustomActions.kt]: JSONArray is fragile, i am wondering if we can gson instead. add gson as a dependency and lets use that please.

File: android/app/src/main/java/com/jossephus/chuchu/ui/screens/Settings/TerminalAccessorySettings.kt [hunk 3]
```diff
@@ -306,6 +312,77 @@ internal fun TerminalSettings(
...
+                                    // Keep each key bound to a single command: clear any
+                                    // other command currently using this key.
+                                    newShortcuts.entries
+                                        .filter { it.value == key && it.key != command.id }
+                                        .forEach { newShortcuts[it.key] = "" }
+                                }
+                                newShortcuts[command.id] = key
+                                onBuiltinShortcutsChanged(newShortcuts)
+                            },
+                            label = command.label,
```
Comment (jossephus) [inline android/app/src/main/java/com/jossephus/chuchu/ui/screens/Settings/TerminalAccessorySettings.kt]: lets follow the existing pattern in this screen and show this in bottom sheet please. it feels off right now because its taking a lot of space.

File: android/app/src/main/java/com/jossephus/chuchu/ui/screens/Terminal/TerminalScreen.kt [hunk 4]
```diff
@@ -299,31 +340,76 @@ fun TerminalScreen(
...
+                BuiltinCommand.NewTab to {
+                    vm.duplicateActiveTab()
+                    vm.selectConnectionTab(ConnectionTab.Terminal)
+                    showTabSheet = false
+                },
+                BuiltinCommand.Close to {
+                    val activeId = vm.activeTabId.value
+                    if (activeId != null) vm.closeTab(activeId)
+                },
+                BuiltinCommand.Actions to { settingsRepo.setShowCustomActionsFab(!showCustomActionsFab) },
```
Comment (jossephus) [inline android/app/src/main/java/com/jossephus/chuchu/ui/screens/Terminal/TerminalScreen.kt]: can we move the logic here to ChuchuKeyBindings? I think it makes sense that way.

**Review by jossephus — CHANGES_REQUESTED** [review]

**Comment by jossephus** [comment]
Hey @bhoffman20 I loved the changes introduced here and i am happy i got another person who heavily uses the command key lol.

File: android/app/src/main/java/com/jossephus/chuchu/ui/terminal/TerminalCustomActions.kt (hunk 2 above)
Comment (DevSrSouza) [reply android/app/src/main/java/com/jossephus/chuchu/ui/terminal/TerminalCustomActions.kt]: Use kotlinx serialization please, is type safer and kotlin standard

Comment (bhoffman20) [reply android/app/src/main/java/com/jossephus/chuchu/ui/terminal/TerminalCustomActions.kt]: Went with kotlinx.serialization rather than gson, following @DevSrSouza's suggestion below — type-safe, and no extra runtime dependency beyond the Kotlin plugin. Both `TerminalCustomActionStore` and `BuiltinShortcutStore` use it now, and no `JSONArray`/`org.json` usage remains.

File: android/app/src/main/java/com/jossephus/chuchu/ui/terminal/BuiltinCommands.kt [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        BuiltinCommand.Close.id to "q",
+    )
```
[the remark this answers is not in the document]
Comment (bhoffman20) [reply android/app/src/main/java/com/jossephus/chuchu/ui/terminal/BuiltinCommands.kt]: Not a real conflict — `n` is the intended default for `new tab`. The `r: new tab` visible in the PR screenshot is my own customized binding made via the new settings screen, not the shipped default (called out in the PR description). Leaving the default as `n`.

**Comment by jossephus** [comment]
Thanks @bhoffman20
